### PR TITLE
feat(examples): --profile support across all ezpz.examples.* via shared helpers

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Adds `--profile` support (PyTorch profiler + pyinstrument) to **all** `ezpz.examples.*` modules. Previously only `ezpz.examples.test` had real profiling; every other example had a long-commented `# and not config.pytorch_profiler` stub and nothing else.

The wiring is centralized in **two shared helpers** so the surface can't drift:

- **`ezpz.cli.flags.add_profiling_args(parser)`** — the 13 profiler flags (`-p`/`--profile`, `--pyinstrument-profiler`, `--rank-zero-only`, `--pytorch-profiler-{wait,warmup,active,repeat}`, `--profile-memory`/`--record-shapes`/`--with-stack`/`--with-flops`/`--with-modules` as `BooleanOptionalAction`, `--acc-events`). Extracted from the block that was inlined in `build_test_parser` (which now calls it; `--save-datasets` stays test-specific).
- **`ezpz.profile.profiling_context_from_args(args, outdir)`** — namespace→context-manager adapter over the existing `get_profiling_context`. Returns `nullcontext()` (yielding `None`) when no profile flag is set, so **default runs are completely unaffected**. Sets `strict=not pyinstrument` so `--pyinstrument-profiler` actually activates (it was a dead flag before — only `PYINSTRUMENT_PROFILER=1` worked), and still honors that env var for backward compat.

## Per-example wiring

| Module | How |
|---|---|
| `fsdp_tp`, `vit`, `diffusion` | `add_profiling_args` on the parser; `profiler.step()` after `optimizer.step()`; loop wrapped in `profiling_context_from_args` |
| `fsdp` | context created in `fsdp_main`, threaded into `train()` across the split epoch/batch loop so the schedule spans the whole run |
| `hf_trainer` | HF Trainer owns its loop, so: new `HfProfileArguments` dataclass (`configs.py`, parsed as a 4th `HfArgumentParser` dataclass) + a `ProfilerCallback(TrainerCallback)` (`on_train_begin` enters / `on_step_end` steps / `on_train_end` flushes), registered only when a profile flag is set |
| `test` | routed through the shared helper (one wiring path; also un-breaks its `--pyinstrument-profiler`) |
| `minimal` | **intentionally skipped** (env-var config, no argparse) — documented profiler-free in its docstring |

Examples that report metrics also disable `distributed_history` under `--profile`, so the per-rank metric all-gather doesn't perturb the step times being measured.

Also folds in two small `fsdp_tp` items that were in the working tree from the same investigation:
- **tokens/sec metric**: `train/tps` (global) + `train/tps_per_gpu` (per-rank, torchtitan's `tgs`).
- **debug escape hatches**: `EZPZ_FSDP_GRAD_DIV=0` and `EZPZ_REDUCE_DTYPE=bf16` (both default to current behavior).

## Verification

- All 10 touched files byte-compile; `add_profiling_args` / `profiling_context_from_args` / `HfProfileArguments` unit-tested via the repo `.venv` (defaults match, flag toggles work, `nullcontext` default + env-var path confirmed).
- `--help` shows exactly 13 profile flags on `fsdp_tp` / `fsdp` / `vit` / `diffusion`.
- **End-to-end live run** (Sunspot, 24×PVC, 2 nodes, `rc=0`): the user's target config
  ```
  ezpz launch python3 -m ezpz.examples.fsdp_tp --model=xl --hf-limit=1000 \
    --seq-len=8192 --batch-size=2 --tokenizer_name=google/gemma-7b --tp=1 \
    --sharding-strategy=hybrid_shard --compile-mode=max-autotune --compile \
    --loss-impl=compiled --act-mem-budget=0.25 --profile --rank-zero-only
  ```
  parsed the flags cleanly and wrote 4 chrome traces (schedule `wait=1,warmup=2,active=3,repeat=5`).

### Profiling finding (xl / seq=8192 / bs=2, compiled loss)

MFU ~19.7%, ~58.8 TFLOP/s/gpu, 4.04 s/step (fwd 0.70s / **bwd 3.33s**). Device-kernel self-time: **attention backward 53%**, GEMMs (lm_head+FFN+qkv) 22.7%, attention fwd 5.9%, **cross-entropy only 1.6%** (compiled CE collapsed the vocab-loss sink). At full depth the bottleneck is the vendor cutlass-sycl FA **backward** kernel — a shift from the earlier single-layer eager profile where the 256K-vocab lm_head/loss dominated.

## Notes

- No behavior change to existing runs (no flag ⇒ `nullcontext`).
- Flag spellings match `test`'s exactly, so docs/scripts/muscle-memory carry over.

## Summary by Sourcery

Introduce shared profiling CLI helpers and wire consistent profiler support across all training examples and the HuggingFace Trainer.

New Features:
- Add a shared add_profiling_args helper to expose a unified profiler flag set across ezpz.examples.*.
- Provide a profiling_context_from_args adapter to translate CLI flags into an optional profiling context manager used around training loops.
- Enable configurable profiling for the HuggingFace Trainer example via HfProfileArguments and a ProfilerCallback driving the profiler lifecycle.
- Add tokens-per-second throughput metrics to the fsdp_tp example output.

Bug Fixes:
- Fix previously non-functional pyinstrument profiler activation by making the CLI flag properly opt-in while retaining env-var support.
- Restore and standardize profiling behavior in the test example so pyinstrument and torch profiler paths work consistently.

Enhancements:
- Refactor the test example to use the shared profiling helpers while preserving its existing CLI surface and behavior.
- Update fsdp, fsdp_tp, vit, and diffusion examples to integrate the shared profiler flags and step the profiler once per optimizer step, without affecting default runs.
- Disable distributed metric history aggregation while profiling in relevant examples to avoid perturbing measured step times.
- Introduce environment-controlled escape hatches for FSDP gradient division and reduce-scatter dtype configuration in fsdp_tp.
- Ensure HuggingFace Trainer argument parsing cleanly separates model, data, training, and profiling concerns via dedicated dataclasses.

### CODE REVIEW

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -1686,9 +1707,20 @@ def train(
...
+        # a large-vocab output projection (e.g. agpt's 256K) the single
+        # reduce-scatter tensor can exceed CCL's ~2GB-per-message MPI limit
+        # (256K*2048*4B = 2.1GB) → `atl_mpi !req.is_completed`. Set
+        # EZPZ_REDUCE_DTYPE=bf16 to halve the collective size (1.05GB) and
+        # stay under the limit.
+        _reduce_dtype = (
+            torch.bfloat16
+            if os.environ.get("EZPZ_REDUCE_DTYPE", "fp32").lower() in ("bf16", "bfloat16")
+            else torch.float32
+        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Done in 40ae2a3 — `EZPZ_REDUCE_DTYPE` is now resolved against {`fp32`,`bf16`,`bfloat16`} (case-insensitive) and raises `ValueError` on anything else. The silent fp32 fallback was especially bad here because it would re-trigger the exact CCL >2GB reduce-scatter failure the bf16 hatch exists to avoid.

File: src/ezpz/examples/hf_trainer.py [hunk 2]
```diff
@@ -60,11 +60,75 @@ def on_log(self, args, state, control, logs=None, **kwargs):
...
+    HfDataTrainingArguments,
+    HfModelArguments,
+    HfProfileArguments,
+)
+from ezpz.profile import get_profiling_context
 
 logger = ezpz.get_logger(__name__)
 
 
+class ProfilerCallback(TrainerCallback):
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/hf_trainer.py:73]: Done in 9f4b22b — `ProfilerCallback.on_train_begin` now calls `profiling_context_from_args(self._args, outdir=...)` directly, dropping the manual field mapping. `HfProfileArguments` exposes the same field names the adapter reads, so torch/pyinstrument selection + strictness + rank-zero all come from the shared path now; the callback only owns the Trainer lifecycle. This also dissolved the confusing `strict=self._args.pytorch_profiler` line you flagged in the high-level notes.

**Comment by saforem2** [comment]
Thanks @sourcery-ai — addressed all three high-level points in code:

1. **`HfProfileArguments.rank_zero_only` default** — aligned to `False` to match `add_profiling_args` (`--rank-zero-only`), so HF profiling no longer diverges from the other entrypoints (9f4b22b).
2. **`distributed_history` only gated on `pytorch_profiler`** — correct, `--pyinstrument-profiler` runs were still doing the cross-rank metric all-gather. Now gated on *either* profiler in fsdp_tp/fsdp/diffusion/test (40ae2a3).
3. **`strict=self._args.pytorch_profiler` reading backwards** — dissolved: `ProfilerCallback` now delegates to `profiling_context_from_args`, which owns the `strict=not pyinstrument` logic in one place, so the callback no longer sets `strict` at all (9f4b22b).

All four CI checks were green before these fixes; re-running now.

File: src/ezpz/configs.py [hunk 3]
```diff
@@ -964,6 +964,73 @@ def __post_init__(self):
...
+    ``HfArgumentParser`` renders ``bool`` fields as a single store-true
+    flag). Field names match the argparse spellings used by the other
+    ``ezpz.examples.*`` modules so semantics carry over; consumed by
+    ``ezpz.examples.hf_trainer.ProfilerCallback`` via
+    :func:`ezpz.profile.get_profiling_context`.
+    """
+
+    pytorch_profiler: bool = field(
+        default=False,
+        metadata={"help": "Enable the torch profiler (chrome traces)."},
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/configs.py]: Fixed in 354f14e — added `metadata['aliases']=['--profile','-p']` to `HfProfileArguments.pytorch_profiler`. transformers 5.12.1's HfArgumentParser supports aliases (`aliases = kwargs.pop('aliases', [])`), so `--profile`/`-p` now reach `pytorch_profiler` for hf_trainer too, matching the shared surface. Verified by a new test (`test_profile_alias_exposed`).

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -2136,6 +2176,16 @@ def train(
...
+            # Throughput. `--batch-size` is the PER-DP-RANK micro-batch, so
+            # tokens processed per rank per step is batch_size * seq_len.
+            #   - train/tps_per_gpu  : per-rank tokens/sec (torchtitan's `tgs`)
+            #   - train/tps          : global tokens/sec across all DP ranks
+            # (dp ranks = world_size / tp; tp ranks process the same tokens).
+            if dt_step > 0:
+                tokens_per_rank = args.batch_size * local_seq_len
+                metrics["train/tps_per_gpu"] = tokens_per_rank / dt_step
+                dp_size = max(1, world_size // args.tp)
+                metrics["train/tps"] = tokens_per_rank * dp_size / dt_step
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Good catch — fixed in e7b593e. Under SP each TP rank holds a distinct seq shard (labels sliced via `_slice_for_sequence_parallel`) and `local_seq_len` is the per-shard length, so `* dp_size` undercounted by ~tp. Now `tokens_per_rank * world_size`; identical at tp=1.

File: src/ezpz/profile.py [hunk 5]
```diff
@@ -167,6 +167,76 @@ def trace_handler(p: torch.profiler.profile):
...
+    """
+    pyinstrument = bool(getattr(args, "pyinstrument_profiler", False))
+    pytorch = bool(getattr(args, "pytorch_profiler", False))
+    # Honor the legacy env-var opt-in so callers that relied on
+    # PYINSTRUMENT_PROFILER=1 (no flag) keep working.
+    if os.environ.get("PYINSTRUMENT_PROFILER") is not None:
+        pyinstrument = True
+    if not (pyinstrument or pytorch):
+        return nullcontext()
+    return get_profiling_context(
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/profile.py:237]: Fixed in 1e90bd8 — the pyinstrument branch now forwards `outdir` + `rank_zero_only` to `get_context_manager`, and `PyInstrumentProfiler` was also hardcoding its outdir to cwd (ignoring the arg), which is now fixed too. So `profiling_context_from_args(args, outdir)` controls where pyinstrument profiles land. Default (cwd) unchanged when no outdir given.

File: src/ezpz/examples/fsdp.py [hunk 6]
```diff
@@ -425,38 +437,47 @@ def fsdp_main(args: argparse.Namespace) -> None:
...
         project_name=WBPROJ_NAME,
         config={"args": vars(args), **ezpz.get_dist_info()},
+        # Disable cross-rank history aggregation while profiling — the
+        # all-gather of per-rank metrics perturbs the very step times the
+        # profiler is measuring.
         distributed_history=(
-            1 < ezpz.get_world_size() <= 384  # and not config.pytorch_profiler
+            1 < ezpz.get_world_size() <= 384
+            and not getattr(args, "pytorch_profiler", False)
         ),
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp.py:447]: Fixed in 40ae2a3 — `distributed_history` now also checks `pyinstrument_profiler`, so either profiler disables the cross-rank all-gather.

File: src/ezpz/examples/diffusion.py [hunk 7]
```diff
@@ -498,8 +506,12 @@ def train(
...
         jsonl_path=metrics_path,
         jsonl_overwrite=True,
+        # Disable cross-rank history aggregation while profiling — the
+        # all-gather of per-rank metrics perturbs the very step times the
+        # profiler is measuring.
         distributed_history=(
-            1 < ezpz.get_world_size() <= 384  # and not config.pytorch_profiler
+            1 < ezpz.get_world_size() <= 384
+            and not getattr(args, "pytorch_profiler", False)
         ),
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/diffusion.py:516]: Fixed in 40ae2a3 — now gated on either profiler (`pytorch_profiler` or `pyinstrument_profiler`).

File: src/ezpz/examples/fsdp_tp.py [hunk 8]
```diff
@@ -1955,8 +1987,12 @@ def train(
...
         jsonl_path=metrics_path,
         jsonl_overwrite=True,
+        # Disable cross-rank history aggregation while profiling — the
+        # all-gather of per-rank metrics perturbs the very step times the
+        # profiler is measuring.
         distributed_history=(
-            1 < world_size <= 384  # and not config.pytorch_profiler
+            1 < world_size <= 384
+            and not getattr(args, "pytorch_profiler", False)
         ),
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2005]: Fixed in 40ae2a3 — now gated on either profiler.

File: src/ezpz/examples/hf_trainer.py [hunk 9]
```diff
@@ -60,11 +60,75 @@ def on_log(self, args, state, control, logs=None, **kwargs):
...
+            with_modules=self._args.with_modules,
+            acc_events=self._args.acc_events,
+            profile_memory=self._args.profile_memory,
+            outdir=self._outdir,
+            # pyinstrument opt-in is the flag itself, not an env var.
+            strict=self._args.pytorch_profiler,
+        )
+        # __enter__ returns the torch profiler (has .step()) or the
+        # pyinstrument wrapper / nullcontext target (None) — store it.
+        self._profiler = self._cm.__enter__()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/hf_trainer.py:104]: Fixed in 9f4b22b — `ProfilerCallback` now delegates to `profiling_context_from_args`, so the type/strictness/rank-zero logic lives in one place and HF picks up shared fixes (e.g. the pyinstrument-outdir fix above) automatically.

File: src/ezpz/profile.py [hunk 10]
```diff
@@ -167,6 +167,76 @@ def trace_handler(p: torch.profiler.profile):
     )
 
 
+def profiling_context_from_args(
+    args: Any,
+    outdir: Optional[str | Path | os.PathLike] = None,
+) -> AbstractContextManager:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/profile.py:193]: Added in 490bc42 — `TestProfilingContextFromArgs` in tests/test_profile.py covers nullcontext-when-no-flag, flag activation, the PYINSTRUMENT_PROFILER env-var opt-in, and the missing-attrs default.

File: src/ezpz/cli/flags.py [hunk 11]
```diff
@@ -48,62 +48,33 @@ def _spare_nodes_value(value: str) -> "int | str":
...
+
+    - ``--pyinstrument-profiler`` — statistical (pyinstrument) profiler.
+    - ``-p`` / ``--profile`` (dest ``pytorch_profiler``) — torch profiler.
+    - ``--rank-zero-only`` — only profile rank 0 (one trace, not N).
+    - ``--pytorch-profiler-{wait,warmup,active,repeat}`` — schedule.
+    - ``--profile-memory`` / ``--record-shapes`` / ``--with-stack`` /
+      ``--with-flops`` / ``--with-modules`` — torch profiler detail
+      toggles (``BooleanOptionalAction`` → each has a ``--no-*`` form).
+    - ``--acc-events`` — accumulate events across schedule cycles.
+    """
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/cli/flags.py:77]: Added in 490bc42 — `tests/test_flags_profiling.py::TestAddProfilingArgs` pins the full flag set, defaults, the `--no-*` BooleanOptionalAction forms, and `-p`/`--profile`.

File: src/ezpz/configs.py [hunk 12]
```diff
@@ -964,6 +964,73 @@ def __post_init__(self):
...
+class HfProfileArguments:
+    """Profiler controls for the HuggingFace Trainer example.
+
+    Mirrors the subset of :func:`ezpz.cli.flags.add_profiling_args` that
+    applies to a Trainer-owned loop (no `--no-*` toggles — HF's
+    ``HfArgumentParser`` renders ``bool`` fields as a single store-true
+    flag). Field names match the argparse spellings used by the other
+    ``ezpz.examples.*`` modules so semantics carry over; consumed by
+    ``ezpz.examples.hf_trainer.ProfilerCallback`` via
+    :func:`ezpz.profile.get_profiling_context`.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/configs.py:977]: Added in 490bc42 — `TestHfProfileArguments` covers the defaults (incl. `rank_zero_only=False` matching the CLI) and the `-p`/`--profile` aliases.
